geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/index/AttributeIndex.scala

```diff
@@ -0,0 +1,355 @@
+import org.locationtech.geomesa.index.index.AttributeIndex
+import org.locationtech.geomesa.utils.geotools.RichAttributeDescriptors.RichAttributeDescriptor
+import org.locationtech.geomesa.utils.index.{IndexMode, VisibilityLevel}
+import org.locationtech.geomesa.utils.stats.IndexCoverage.IndexCoverage
+import org.locationtech.geomesa.utils.stats.{Cardinality, IndexCoverage, Stat}
+import org.opengis.feature.simple.SimpleFeatureType
+import org.opengis.filter.Filter
+
+import scala.util.Try
+
+// secondary z-index
+case object AttributeIndex extends AccumuloFeatureIndex with AccumuloIndexAdapter
+    with AttributeIndex[AccumuloDataStore, AccumuloFeature, Mutation, Range] with AttributeSplittable {
+
+  import scala.collection.JavaConversions._
+
+  type ScanPlanFn = (SimpleFeatureType, Option[Filter], Option[(String, SimpleFeatureType)]) => BatchScanPlan
+
+  override val version: Int = 4
+
+  override val serializedWithId: Boolean = false
+
+  override val hasPrecomputedBins: Boolean = false
+
+  override def configureSplits(sft: SimpleFeatureType, ds: AccumuloDataStore): Unit = {
+    import scala.collection.JavaConversions._
+    val table = getTableName(sft.getTypeName, ds)
+    val splits = getSplits(sft).map(new Text(_)).toSet -- ds.tableOps.listSplits(table)
+    if (splits.nonEmpty) {
+      ds.tableOps.addSplits(table, ImmutableSortedSet.copyOf(splits.toArray))
+    }
+  }
+
+  override def writer(sft: SimpleFeatureType, ds: AccumuloDataStore): (AccumuloFeature) => Seq[Mutation] = {
+    val getRows = getRowKeys(sft)
+    val coverages = sft.getAttributeDescriptors.map(_.getIndexCoverage()).toArray
+    (wf) => getRows(wf).map { case (i, r) => createInsert(r, wf, coverages(i)) }
+  }
+
+  override def remover(sft: SimpleFeatureType, ds: AccumuloDataStore): (AccumuloFeature) => Seq[Mutation] = {
+    val getRows = getRowKeys(sft)
+    val coverages = sft.getAttributeDescriptors.map(_.getIndexCoverage()).toArray
+    (wf) => getRows(wf).map { case (i, r) => createDelete(r, wf, coverages(i)) }
+  }
+
+  protected def createInsert(row: Array[Byte], feature: AccumuloFeature, coverage: IndexCoverage): Mutation = {
+    val mutation = new Mutation(row)
+    val values = coverage match {
+      case IndexCoverage.FULL => feature.fullValues
+      case IndexCoverage.JOIN => feature.indexValues
+    }
+    values.foreach(v => mutation.put(v.cf, v.cq, v.vis, v.value))
+    mutation
+  }
+
+  protected def createDelete(row: Array[Byte], feature: AccumuloFeature, coverage: IndexCoverage): Mutation = {
+    val mutation = new Mutation(row)
+    val values = coverage match {
+      case IndexCoverage.FULL => feature.fullValues
+      case IndexCoverage.JOIN => feature.indexValues
+    }
+    values.foreach(v => mutation.putDelete(v.cf, v.cq, v.vis))
+    mutation
+  }
+
+  override protected def createInsert(row: Array[Byte], feature: AccumuloFeature): Mutation =
+    throw new NotImplementedError("Should be using enhanced version with IndexCoverage")
```

Comment:
that it should be using the method 20 lines up. this method is part of the generic `GeoMesaFeatureIndex` trait, but it can't account for join/full indices. we've overridden the `writer` and `remover` methods so that it doesn't get called, but it's still part of the API so we can't remove it - this is just a sanity check.